### PR TITLE
Add timers command

### PR DESCRIPTION
**Pull request by lachupe**

### CODE REVIEW

File: src/parser/mumexmlparser.cpp [hunk 1]
```diff
@@ -172,23 +176,36 @@ void MumeXmlParser::parse(const TelnetData &data, const bool isGoAhead)
...
+{
+    // lvl100 hungry/thirsty?
+    // shall this be simplified somehow?
+    static QRegExp statExp("Needed: * Alert: *.", Qt::CaseSensitive, QRegExp::Wildcard);
+    static QRegExp statExpCond("Needed: * Alert: *. Cond*", Qt::CaseSensitive, QRegExp::Wildcard);
+    static QRegExp lvl100statExp("Gold: *. Alert: *.", Qt::CaseSensitive, QRegExp::Wildcard);
+
+    if (statExp.exactMatch(currentString) || lvl100statExp.exactMatch(currentString)
+        || statExpCond.exactMatch(currentString)) {
+        sendToUser(m_timers.getStatCommandEntry(), isGoAhead);
```
Comment (nschimme) [inline src/parser/mumexmlparser.cpp:208]: isGoAhead is only for prompts so we don't need to pass this argument through. Given this, we can also keep this logic confined to parseMudCommands()

File: src/timers/CTimers.cpp [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+CTimers::CTimers(QObject *parent)
+    : QObject(parent)
+{
+    m_nextId = 1;
+}
+
+CTimers::~CTimers() {}
+
+void CTimers::addTimer(QByteArray name, QByteArray desc)
```
Comment (nschimme) [inline src/timers/CTimers.cpp:18]: We can avoid making copies of the arguments here.

File: src/timers/CTimers.cpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+    // append the shot-down event pending
+    QTimer::singleShot(time, this, SLOT(finishCountdownTimer()));
+
+    m_countdowns.push_back(countdown);
+}
+
+void CTimers::finishCountdownTimer()
+{
+    QMutexLocker locker(&m_lock);
```
Comment (nschimme) [inline src/timers/CTimers.cpp:79]: I'm wondering if a mutex is even necessary here. Isn't the Thread owned by the same object as the Proxy and Parser so it's the same thread?

File: src/timers/CTimers.cpp [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+    int deleted = 0;
```
Comment (nschimme) [inline src/timers/CTimers.cpp:81]: Does it matter how many we delete? Can this be a boolean?

File: src/proxy/proxy.cpp [hunk 5]
```diff
@@ -121,10 +121,12 @@ void Proxy::slot_start()
     m_mudTelnet = makeQPointer<MudTelnet>(this);
     m_telnetFilter = makeQPointer<TelnetFilter>(this);
     m_mpiFilter = makeQPointer<MpiFilter>(this);
+    m_timers = makeQPointer<CTimers>(this);
```
Comment (nschimme) [inline src/proxy/proxy.cpp:124]: Why do we create this object in the Proxy as opposed to in the AbstractParser?

File: src/parser/mumexmlparser.cpp [hunk 6]
```diff
@@ -160,23 +160,34 @@ void MumeXmlParser::parse(const TelnetData &data, const bool isGoAhead)
...
-                && tempStr.at(1) == m_snoopChar.value() && tempStr.at(2) == ' ') {
-                // Remove snoop prefix (i.e. "&J Exits: north.")
-                tempStr = tempStr.mid(3);
-            }
-            parseMudCommands(tempStr);
-        }
+void MumeXmlParser::printTimersExtensions(const QString currentString, bool isGoAhead) {
+    // lvl100 hungry/thirsty?
+    // shall this be simplified somehow?
+    static QRegExp statExp("Needed: * Alert: *.", Qt::CaseSensitive, QRegExp::Wildcard);
```
[the remark this answers is not in the document]
Comment (lachupe) [reply src/parser/mumexmlparser.cpp:202]: Nice, I'll try that one!

File: src/parser/mumexmlparser.cpp (hunk 1 above)
Comment (lachupe) [reply src/parser/mumexmlparser.cpp:208]: Not yet sure I understood the hint about keeping the logic in parseMudCommands (need to take a look at the code). But I guess i can remove the isGoAhead. This is for sure only about injecting additional lines right after the matching one.

File: src/timers/CTimers.cpp (hunk 3 above)
Comment (lachupe) [reply src/timers/CTimers.cpp:79]: Well, I wasn't sure. 
At least 3 things may happen (in theory) at the same time:
- timer for the countdown kicking in and tweaking the countdowns data
- User input adds new entries or remove the old ones
- Proxy (if i recall that right) attempting to retrieve the entries

Now, I'm not sure if this is really in the same thread or not, so I kept the same protection as the old code in PandoraMapper.

File: src/timers/CTimers.cpp (hunk 4 above)
Comment (lachupe) [reply src/timers/CTimers.cpp:81]: Probably I'd better opt for writing out how many items were deleted. timers may have the same "name", so multiples may be deleted.

File: src/timers/CTimers.cpp [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+{
+    if (m_timers.size() == 0)
+        return "";
+
+    QString line = "Timers:\r\n";
+    for (std::list<TTimer>::iterator it = m_timers.begin(); it != m_timers.end(); ++it) {
+        line += QString("- %1 < %2 > (up for - %3)\r\n")
+                    .arg(it->name.data(), it->desc.data(), msToMinSec(it->timer.elapsed()));
+    }
+    return line.toLocal8Bit();
```
[the remark this answers is not in the document]
Comment (lachupe) [reply src/timers/CTimers.cpp:121]: Sounds good to me. I don't think there is any solid reason (matching and regexps may be?) to keep QStrings and QByteArray. This is also only due to historical reasons of the code coming from pandora.

File: src/timers/CTimers.cpp (hunk 2 above)
Comment (lachupe) [reply src/timers/CTimers.cpp:18]: :+1:

File: src/timers/CTimers.h [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+#include <QMutex>
+#include <QObject>
+
+struct TTimer
+{
+    int id;
+    QByteArray name;
+    QByteArray desc;
+    int duration;
+    QElapsedTimer timer;
```
[the remark this answers is not in the document]
Comment (nschimme) [reply src/timers/CTimers.h:16]: I couldn't think of one. I was wondering if we need to track N different QTimers here of we just had a "master" one somewhere that would parse the data and learn when to run the next callback. This thing here is more of a thought experiment than a real need so feel free to ignore it.

File: src/parser/mumexmlparser.cpp (hunk 1 above)
Comment (nschimme) [reply src/parser/mumexmlparser.cpp:208]: Ah, let me give the backstory. The general Parser file is getting huge so I'm trying to isolate similar logic to other files

File: src/parser/mumexmlparser.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (lachupe) [reply src/parser/mumexmlparser.cpp]: Works! Regexp magic! ;)

File: src/timers/CTimers.cpp (hunk 2 above)
Comment (lachupe) [reply src/timers/CTimers.cpp]: I switched to std::strings and using const references. 
Saw one issue with latin-1 encoding, but that got to be my conversions somewhere...

File: src/timers/CTimers.cpp (hunk 4 above)
Comment (lachupe) [reply src/timers/CTimers.cpp]: Yeah, replaced with a boolean. 
Int wasn't necessary at all.

File: src/timers/CTimers.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (lachupe) [reply src/timers/CTimers.cpp]: Replaced! Checking for bugs.

File: src/parser/mumexmlparser.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (nschimme) [reply src/parser/mumexmlparser.cpp]: I just realized it won't work for ANSI colors. I'll have to give you and updated regex.

File: src/timers/CTimers.cpp [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        } else {
+            ++it;
+        }
+    }
+
+    // but still, at least 1ms
+    delta = std::max(delta, 1);
+
+    // as of Qt 5.2.1 singleShot timers are not precise enough and might, easily, fire earlier!
+    // Aza: check? I have no idea what this is about anymore
```
Comment (nschimme) [inline src/timers/CTimers.cpp:123]: Should we re-check rather than carry this comment forward?

File: src/parser/mumexmlparser.h [hunk 10]
```diff
@@ -99,6 +100,7 @@ class MumeXmlParser final : public AbstractParser
     NODISCARD bool element(const QByteArray &);
     void move();
     NODISCARD std::string snoopToUser(const std::string_view str);
+    void printTimersExtensions(const QString currentString);
```
Comment (nschimme) [inline src/parser/mumexmlparser.h:103]: can pass by const reference

File: src/CMakeLists.txt [hunk 11]
```diff
@@ -386,6 +387,8 @@ set(mmapper_SRCS
     syntax/User.h
     syntax/Value.cpp
     syntax/Value.h
+    timers/CTimers.cpp
```
Comment (nschimme) [inline src/CMakeLists.txt:390]: Can we remove the C prefix? It doesn't match the style of the rest of the files

File: src/parser/mumexmlparser.cpp [hunk 12]
```diff
@@ -172,23 +176,31 @@ void MumeXmlParser::parse(const TelnetData &data, const bool isGoAhead)
...
-            if (m_snoopChar.has_value() && tempStr.length() > 3 && tempStr.at(0) == '&'
-                && tempStr.at(1) == m_snoopChar.value() && tempStr.at(2) == ' ') {
-                // Remove snoop prefix (i.e. "&J Exits: north.")
-                tempStr = tempStr.mid(3);
-            }
-            parseMudCommands(tempStr);
-        }
+void MumeXmlParser::printTimersExtensions(const QString currentString)
+{
+    static QRegExp statLineExp("^(?:Needed:(?: [\\d,]+ xp)?(?:,? [\\d,]+ tp)\\. )?Gold: [\\d,]+\\.(?: Iv: [^.]+\\.)? Alert: \\w+\\.(?: Condition: [^.]+\\.)?");
```
Comment (nschimme) [inline src/parser/mumexmlparser.cpp:200]: Oh, `QRegExp` is a deprecated class in Qt. Let's switch to `QRegularExpression`

File: src/timers/CTimers.cpp [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+
+void CTimers::addCountdown(const std::string &name, const std::string &desc, int time)
+{
+    QMutexLocker locker(&m_lock);
+
+    TTimer countdown;
+    countdown.name = name;
+    countdown.desc = desc;
+    countdown.timer.start();
+    countdown.id = m_nextId++;
```
Comment (nschimme) [inline src/timers/CTimers.cpp:84]: Do we even need the `id` ?

File: src/timers/CTimers.cpp (hunk 3 above)
Comment (nschimme) [reply src/timers/CTimers.cpp:115]: I left it

File: src/timers/CTimers.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (nschimme) [reply src/timers/CTimers.h]: I found a simpler solution, see the refactor

File: src/parser/mumexmlparser.cpp [hunk 14]
```diff
@@ -170,23 +174,21 @@ void MumeXmlParser::parse(const TelnetData &data, const bool isGoAhead)
         m_tempCharacters.clear();
     }
     if (!m_lineToUser.isEmpty()) {
-        sendToUser(m_lineToUser, isGoAhead);
```
Comment (nschimme) [inline src/parser/mumexmlparser.cpp:173]: This line might cause a regression

**Pull request merged**